### PR TITLE
8313452: Improve Classfile API attributes handling safety

### PR DESCRIPTION
**Pull request by asotona**
This PR improved Classfile API attributes handling safety by introduction of attribute stability indicator
and by extension of UnknownAttributesOption to more general AttributesProcessingOption.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313452](https://bugs.openjdk.org/browse/JDK-8313452): Improve Classfile API attributes handling safety (**Enhancement** - P4)

### Reviewers
 * [Brian Goetz](https://openjdk.org/census#briangoetz) (@briangoetz - **Reviewer**) ⚠️ Review applies to [392ba560](https://git.openjdk.org/jdk/pull/15101/files/392ba56071dfdbc99a7015c20518a9902f9afad3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15101/head:pull/15101` \
`$ git checkout pull/15101`

Update a local copy of the PR: \
`$ git checkout pull/15101` \
`$ git pull https://git.openjdk.org/jdk.git pull/15101/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15101`

View PR using the GUI difftool: \
`$ git pr show -t 15101`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15101.diff">https://git.openjdk.org/jdk/pull/15101.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15101#issuecomment-1706216168)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractDirectBuilder.java [hunk 1]
```diff
@@ -54,7 +54,9 @@ public void setOriginal(M original) {
     }
 
     public void writeAttribute(Attribute<?> a) {
-        if (context.attributesProcessingOption().isAllowed(a.attributeMapper().attributeStability())) {
+        if (a instanceof UnboundAttribute ||
+            5 - a.attributeMapper().attributeStability().ordinal()
+                    > context.attributesProcessingOption().ordinal()) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractDirectBuilder.java:59]: Maybe extract the `5 - a.attributeMapper().attributeStability().ordinal() > context.attributesProcessingOption().ordinal()` check into a helper method, as the current expression is too noisy:
```java
import jdk.internal.classfile.AttributeMapper.AttributeStability;
import jdk.internal.classfile.Classfile.AttributesProcessingOption;

private static final int ATTRIBUTE_STABILITY_COUNT = AttributeStability.values().length;
public static boolean isAttributeAllowed(
	final AttributeStability stability,
	final AttributesProcessingOption processingOption
) {
	return ATTRIBUTE_STABILITY_COUNT - stability.ordinal() > processingOption.ordinal();
}
```

Also, the magic number should probably be replaced with a constant storing the value of `AttributeStability.values().length`.

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractDirectBuilder.java [hunk 2]
```diff
@@ -52,6 +52,8 @@ public void setOriginal(M original) {
     }
 
     public void writeAttribute(Attribute<?> a) {
-        attributes.withAttribute(a);
+        if (Util.isAttributeAllowed(a, context.attributesProcessingOption())) {
+            attributes.withAttribute(a);
+        }
     }
```
Comment (briangoetz) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractDirectBuilder.java:58]: The "is allowed" check should always say yes for unbound attributes, and only apply the option for bound attributes.  (You may already have done that, but I wanted to make sure this was captured in the review.)

File: src/java.base/share/classes/jdk/internal/classfile/impl/Util.java [hunk 3]
```diff
@@ -54,6 +54,15 @@ public class Util {
...
     }
 
+    private static final int ATTRIBUTE_STABILITY_COUNT = AttributeMapper.AttributeStability.values().length;
+
+    public static boolean isAttributeAllowed(final Attribute<?> attr,
+                                             final Classfile.AttributesProcessingOption processingOption) {
+        return attr instanceof BoundAttribute
+                ? ATTRIBUTE_STABILITY_COUNT - attr.attributeMapper().attributeStability().ordinal() > processingOption.ordinal()
+                : true;
+    }
```
Comment (briangoetz) [inline src/java.base/share/classes/jdk/internal/classfile/impl/Util.java:64]: Ignore earlier comment, looks like you've got this :)

**Review by briangoetz — APPROVED** [review]

File: src/java.base/share/classes/jdk/internal/classfile/AttributeMapper.java [hunk 4]
```diff
@@ -75,4 +110,9 @@ default int validSince() {
     default boolean allowMultiple() {
         return false;
     }
+
+    /**
+     * {@return attribute stability indicator}
+     */
+    AttributeStability attributeStability();
```
[the remark this answers is not in the document]
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/AttributeMapper.java]: Renamed, thanks!

**Pull request closed without merging**